README.md

[![Arduino CI](https://github.com/RobTillaart/MCP4725/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/MCP4725/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/MCP4725/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/MCP4725/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/MCP4725/actions/workflows/jsoncheck.yml)
[![GitHub issues](https://img.shields.io/github/issues/RobTillaart/MCP4725.svg)](https://github.com/RobTillaart/MCP4725/issues)

[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/MCP4725/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/MCP4725.svg?maxAge=3600)](https://github.com/RobTillaart/MCP4725/releases)
[![PlatformIO Registry](https://badges.registry.platformio.org/packages/robtillaart/library/MCP4725.svg)](https://registry.platformio.org/libraries/robtillaart/MCP4725)


# MCP4725

Arduino library for 12 bit I2C DAC - MCP4725.


## Description

The MCP4725 is an I2C single channel 12 bit Digital to Analog Converter (DAC). 
It is possible to have up to 8 MCP4725 on one I2C bus. 
The MCP4725 supports 100 KHz 400 KHz and 3.4 MHz bus speeds.

The output of the MCP4725 depends on the voltage supplied, which is in the range 
of 2.7V .. 5.5V. Check datasheet for the details.


#### 0.4.0 Breaking change

Version 0.4.0 introduced a breaking change.
You cannot set the pins in **begin()** any more.
This reduces the dependency of processor dependent Wire implementations.
The user has to call **Wire.begin()** and can optionally set the Wire pins 
before calling **begin()**.


#### Related

- https://github.com/RobTillaart/AD56x8 (multi channel)
- https://github.com/RobTillaart/AD568X (single channel lower resolution)
- https://github.com/RobTillaart/MCP_DAC (SPI interface)


## Interface

```cpp
#include "MCP4725.h"
```

### Constructor

- **MCP4725(uint8_t deviceAddress, TwoWire \*wire = &Wire)** Constructor, needs I2C address, optional set Wire bus
- **bool begin()** initializes internals.
Returns false if address out of range.
Returns true if deviceAddress can be found on the I2C bus.
- **bool isConnected()** returns true if device (address) can be seen on the I2C bus.
- **uint8_t getAddress())** returns address set in constructor.

### Base

- **int setValue(uint16_t value = 0)** value = 0 .. 4095.  
Uses writeFastMode and does not write to EEPROM.
Therefore it does not update the lastWriteEEPROM timestamp.
The default value is 0.  
Returns 0 on success
- **uint16_t getValue()** returns last value set from cache, this is much faster than readDAC().  
This latter gives the real value from the MCP4725. 
Note: a difference can be caused by power outage a reset etc.
- **int setPercentage(float percentage)** percentage = 0..100.0%.
Convenience wrapper around setValue().
- **float getPercentage()** returns percentage. Wrapper around getValue().
- **int writeDAC(value, bool EEPROM = false)** Writes to DAC and conditionally to EEPROM.  
This latter is for startup / reset behaviour. Check datasheet for the detail behaviour.
- **bool ready()** returns true if a new value can be written to the MCP4725.  
Return false if recently was written to EEPROM.
- **uint16_t readDAC()** reads the current value set in the MCP4725.
- **uint16_t readEEPROM()** reads the current value in the EEPROM of the MCP4725.
- **uint32_t getLastWriteEEPROM()** returns time in millis() when last value was written to EEPROM.

Page 3 of datasheet states it takes **25 - 50 ms** before values are written to EEPROM. 
So the exact timing of a sensor can differ quite a lot. 
When writing to EEPROM with **writeDAC()** one should check it is at least 50 ms ago.
If one know the specific timing of a sensor one can tune this or even make it adaptive.  


### Voltage interface

(Since 0.3.8) 
Assumes linear behaviour over 12 bit from 0..4095 == 0 .. maxVoltage.
The default value is 5.0 volt.
Allows sort of calibration e.g. setting maxVoltage to 4.9 Volt.
Furthermore it can be a preferred interface over percentage and raw values.

- **void setMaxVoltage(float v = 5.0)** configures maximum voltage of Vout.
- **float getMaxVoltage()** return set maximum.
- **void setVoltage(float v)** set the DAC to voltage v.
- **float getVoltage()** get the current setting as a voltage

If this behaviour is not precise enough or should be more "complex" the user can 
always define its own set of wrapper functions around **getValue()** and **setValue()**.


## Experimental

Check datasheet for these functions, (not tested enough yet).

- **int writePowerDownMode(uint8_t PDM, bool EEPROM = false)**
- **uint8_t readPowerDownModeEEPROM()**
- **uint8_t readPowerDownModeDAC()**
- **int powerOnReset()**
- **int powerOnWakeUp()**

More investigations needed for:
- Writing to EEPROM, **ready()** and **getLastWriteEEPROM()**  
checking when and how long the sensor blocks needs to be verified in detail in practice. 


## Address Notes

The address of the MCP4725 in the demo sketches is set to 0x62 as that 
was the address of the sensor I had during writing. 
According to the datasheet the address has the following bit pattern: 

address  011000FFU   (so 0x6?)

         F = Factory set bit
         U = User configurable bit

This means you can have at most 8 MCP4725 in one I2C bus (without multiplexing).
You have to specify the factory bits in your order, but not all shops might
provide all versions of the MCP4725.

```
MCP4725A0T-E/CH:  0110 000U   0x60 - 0x61
MCP4725A1T-E/CH:  0110 001U   0x62 - 0x63
MCP4725A2T-E/CH:  0110 010U   0x64 - 0x65
MCP4725A3T-E/CH:  0110 011U   0x66 - 0x67
```

If one need more DAC's one might have a look at the MCP4728
It has 4 channels per chip (no experience / library yet)


#### RP2040 specific

There are different RP2040 board files.
This library works with RP2040 pico - https://github.com/earlephilhower/arduino-pico
Will give an error on setSDA() and setSCL() with the "Arduino MBED OS" version.

See remark PR #21.

- **bool begin(int sda, int scl)** begin communication with the DAC. 
It has the parameter for selecting on which pins the communication should happen. 
Check RP2040 Pinout for compatible pins. 
When Wire1 is used, it needs to be specified in the constructor with "&Wire1"


## Multi devices on one I2C bus

Normal one can have 1 to 8 MCP4725 on a I2C bus. This depends on the right
choice of devices as 2 address bits are hard coded in the device. 
See paragraph 7.2 datasheet + address notes section above.

But one cannot always order the right devices.
Especially breakout boards often have the same address-range.


#### TCA9548 I2C multiplexer

Use an I2C multiplexer to create multiple "I2C channels" which allows then 
up to 8 devices per channel. Selecting the right devices includes setting 
the I2C multiplexer to the right channel to address the right device. 
This implies access is a bit slower and uses more code.

One (TCA9548) multiplexer allows one to control up to 64 MCP4725's.

- https://github.com/RobTillaart/TCA9548

Note that other multiplexers do exist.


#### Use A0 address pin as a SELECT pin

(Experimental)  
Need to do more tests to see how this solution behaves in practice.  
Verified to work - see https://forum.arduino.cc/t/using-digital-pins-to-control-two-mcp4725-modules/1161482/7.

The assumption here is that the devices are all from the same address range (factory bits).

You can control multiple MCP4725 over the hardware I2C bus with an extra IO pin per device.
- Connect the address pin of every MCP4725 to an IO pin which will work as a **SELECT** pin.
- Keep all IO pins **LOW** so the all have effectively the same address == 0x60.
- To select a specific MCP4725 you set the related IO pin to **HIGH** and that one will have address == 0x61.
- Now you can access and control the associated MCP4725 with address 0x61.
- See **mcp4725_multiple.ino** and **mcp4725_multiple_minimal.ino**


## Future

#### Must

- update documentation

#### Should

- test the powerDown modes / functions.
- test A0 (address bit) as SELECT pin.

#### Could

- extend unit tests
- MCP4725_VERSION ==> MCP4725_LIB_VERSION?


## Support

If you appreciate my libraries, you can support the development and maintenance.
Improve the quality of the libraries by providing issues and Pull Requests, or
donate through PayPal or GitHub sponsors.

Thank you,



CHANGELOG.md
# Change Log MCP4725
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.4.0] - 2023-12-08
- refactor API, begin()
- update readme.md
- update examples
- fix **MCP4725_voltage.ino** example
- minor edits

----

## [0.3.9] - 2023-09-25
- add Wire1 support for ESP32

## [0.3.8] - 2023-09-18
- fix #30 Voltage functions (wrapper).
  - **void setMaxVoltage(float v)**
  - **void getMaxVoltage(float v)**
  - **void setVoltage(float v)**
  - **float getVoltage()**
- add examples
- update readme.md

## [0.3.7] - 2023-08-25
- fix #27 - support for multiple MCP4725
  - add two examples.
- update readme.md
- minor edits

## [0.3.6] - 2023-07-19
- fix #25 - buffer overflow in examples
- update keywords.txt
- moved code to MCP4725.cpp
- minor edits

## [0.3.5] - 2023-01-17
- update GitHub actions
- update license
- fix #23 add note about RP2040
- add **MCP4725_MIDPOINT** constant
- minor edits

## [0.3.4] - 2022-10-17
- Add RP2040 support (kudos to intubun)
- Add RP2040 in build-CI
- Add changelog.md

## [0.3.3] - 2021-12-21
- update library.json,
- update license
- minor edits

## [0.3.2] - 2021-06-06
- Verify input of setPercentage()

## [0.3.1] - 2021-05-27
- Fix Arduino-CI / Arduino-lint

## [0.3.0] - 2021-01-15
- Add WireN support (e.g. teensy)

----

## [0.2.3] - 2020-12-26
- Arduino-CI,
- add bool isConnected(),
- update bool begin()

## [0.2.2] - 2020-07-05
- add get/setPercentage();

## [0.2.1] - 2020-07-04
- Add yield();
- add getLastWriteEEPROM();
- update readme.md
- update keywords.txt

## [0.2.0] - 2020-06-20
- use #pragma once;
- remove pre 1.0 support;
- refactor a lot
- rename RDY() -> ready()

----

## [0.1.9] - 2019-10-14
- replace AVR specific TWBR with \_wire->setClock() #131

## [0.1.8] - 2018-10-24
- fix read only variables #115 (kudos to perl1234)

## [0.1.7] - 2017-04-20
- refactor the removed timeout (Thanks to Koepel)

## [0.1.6] - 2017-04-19
- refactor
- remove timeout - https://github.com/RobTillaart/Arduino/issues/63

## [0.1.05] - 2015-03-06
- refactoring, stricter interfaces

## [0.1.04] - 2013-12-04
- improved the generalCall code (still experimental)

## [0.1.03] - 2013-12-01
- added powerDownMode code

## [0.1.02] - 2013-12-01
- added readEEPROM()
- added RDY()

## [0.1.01] - 2013-11-30
- added readDAC()
- added writeDAC(registerwrite)

## [0.1.00] - 2013-11-24
- initial version



test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2020-12-27
// PURPOSE: unit tests for the MCP4725
//          https://github.com/RobTillaart/MCP4725
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// https://github.com/Arduino-CI/arduino_ci/blob/master/cpp/unittest/Assertion.h#L33-L42
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)
// assertNotNull(actual)

#include <ArduinoUnitTests.h>


#include "Arduino.h"
#include "MCP4725.h"


unittest_setup()
{
  fprintf(stderr, "MCP4725_VERSION: %s\n", (char *) MCP4725_VERSION);
}


unittest_teardown()
{
}


//  MCP.begin() has blocking calls
//  cannot be tested without stub
unittest(test_constructor)
{
  MCP4725 MCP(0x62);
  Wire.begin()

  assertEqual(0x62, MCP.getAddress());

  assertEqual(0, MCP.getValue());
  assertEqual(0, MCP.getLastWriteEEPROM());

  fprintf(stderr, "test start\n");
  assureTrue(MCP.isConnected());
  // assertTrue(MCP.begin());
}


unittest(test_invalid_address)
{
  MCP4725 MCP_F(0x22);
  assertFalse(MCP_F.begin());
}


unittest(test_constant)
{
  fprintf(stderr, "test default values\n");
  assertEqual(MCP4725_MAXVALUE, 4095);
  assertEqual(MCP4725_OK, 0);
  assertEqual(MCP4725_VALUE_ERROR, -999);
  assertEqual(MCP4725_REG_ERROR, -998);
  assertEqual(MCP4725_NOT_CONNECTED, -997);

  fprintf(stderr, "test mode values\n");
  assertEqual(MCP4725_PDMODE_NORMAL, 0);
  assertEqual(MCP4725_PDMODE_1K, 1);
  assertEqual(MCP4725_PDMODE_100K, 2);
  assertEqual(MCP4725_PDMODE_500K, 3);

  fprintf(stderr, "other\n");
  assertEqual(MCP4725_MIDPOINT, 2048);
}


unittest(test_get_setValue)
{
  MCP4725 MCP(0x62);

  assertEqual(MCP4725_VALUE_ERROR, MCP.setValue(65535));
  assertEqual(MCP4725_VALUE_ERROR, MCP.setValue(4096));
}


unittest(test_get_setPercentage)
{
  MCP4725 MCP(0x62);

  assertEqual(MCP4725_VALUE_ERROR, MCP.setPercentage(345));
  assertEqual(MCP4725_VALUE_ERROR, MCP.setPercentage(100.1));
}


unittest(test_writeDAC)
{
  MCP4725 MCP(0x62);

  assertEqual(MCP4725_VALUE_ERROR, MCP.writeDAC(4096, false));
  assertEqual(MCP4725_VALUE_ERROR, MCP.writeDAC(4096, true));
}


unittest_main()


//  -- END OF FILE --



MCP4725.cpp
//
//    FILE: MCP4725.cpp
//  AUTHOR: Rob Tillaart
// PURPOSE: Arduino library for 12 bit I2C DAC - MCP4725
// VERSION: 0.4.0
//     URL: https://github.com/RobTillaart/MCP4725


#include "MCP4725.h"


//  registerMode
#define MCP4725_DAC             0x40
#define MCP4725_DACEEPROM       0x60

//  page 22
#define MCP4725_GC_RESET        0x06
#define MCP4725_GC_WAKEUP       0x09


MCP4725::MCP4725(const uint8_t deviceAddress, TwoWire *wire)
{
  _deviceAddress   = deviceAddress;
  _wire            = wire;
  _lastValue       = 0;
  _powerDownMode   = 0;
  _lastWriteEEPROM = 0;
  _maxVoltage      = 5.0;
}


bool MCP4725::begin()
{
  if ((_deviceAddress < 0x60) || (_deviceAddress > 0x67)) return false;
  if (! isConnected()) return false;

  _lastValue = readDAC();
  _powerDownMode = readPowerDownModeDAC();
  return true;
}


bool MCP4725::isConnected()
{
  _wire->beginTransmission(_deviceAddress);
  return (_wire->endTransmission() == 0);
}


uint8_t MCP4725::getAddress()
{
  return _deviceAddress;
}


int MCP4725::setValue(const uint16_t value)
{
  if (value == _lastValue) return 0;
  if (value > MCP4725_MAXVALUE) return MCP4725_VALUE_ERROR;
  int rv = _writeFastMode(value);
  if (rv == 0) _lastValue = value;
  return rv;
}


uint16_t MCP4725::getValue()
{
  return _lastValue;
}


int MCP4725::setPercentage(float percentage)
{
  if ((percentage > 100) || (percentage < 0)) return MCP4725_VALUE_ERROR;
  return setValue(round(percentage * (0.01 * MCP4725_MAXVALUE)));
}


float MCP4725::getPercentage()
{
  return getValue() * (100.0 / MCP4725_MAXVALUE);
}


void MCP4725::setMaxVoltage(float v)
{
  _maxVoltage = v;
}


float MCP4725::getMaxVoltage()
{
  return _maxVoltage;
}


int MCP4725::setVoltage(float v)
{
  return setValue(round((v * MCP4725_MAXVALUE) / _maxVoltage));
}


float MCP4725::getVoltage()
{
  return getValue() * (_maxVoltage / MCP4725_MAXVALUE);
}


//  unfortunately it is not possible to write a different value
//  to the DAC and EEPROM simultaneously or write EEPROM only.
int MCP4725::writeDAC(const uint16_t value, const bool EEPROM)
{
  if (value > MCP4725_MAXVALUE) return MCP4725_VALUE_ERROR;
  while(!ready());
  int rv = _writeRegisterMode(value, EEPROM ? MCP4725_DACEEPROM : MCP4725_DAC);
  if (rv == 0) _lastValue = value;
  return rv;
}


//  ready checks if the last write to EEPROM has been written.
//  until ready all writes to the MCP4725 are ignored!
bool MCP4725::ready()
{
  yield();
  uint8_t buffer[1];
  _readRegister(buffer, 1);
  return ((buffer[0] & 0x80) > 0);
}


uint16_t MCP4725::readDAC()
{
  while(!ready());
  uint8_t buffer[3];
  _readRegister(buffer, 3);
  uint16_t value = buffer[1];
  value = value << 4;
  value = value + (buffer[2] >> 4);
  return value;
}


uint16_t MCP4725::readEEPROM()
{
  while(!ready());
  uint8_t buffer[5];
  _readRegister(buffer, 5);
  uint16_t value = buffer[3] & 0x0F;
  value = value << 8;
  value = value + buffer[4];
  return value;
}


uint32_t MCP4725::getLastWriteEEPROM()
{
  return _lastWriteEEPROM;
};


//  depending on bool EEPROM the value of PDM is written to
//  (false) DAC or
//  (true) DAC & EEPROM,
int MCP4725::writePowerDownMode(const uint8_t PDM, const bool EEPROM)
{
  _powerDownMode = (PDM & 0x03); // mask PDM bits only (written later low level)
  return writeDAC(_lastValue, EEPROM);
}


uint8_t MCP4725::readPowerDownModeEEPROM()
{
  while(!ready());
  uint8_t buffer[4];
  _readRegister(buffer, 4);
  uint8_t value = (buffer[3] >> 5) & 0x03;
  return value;
}


uint8_t MCP4725::readPowerDownModeDAC()
{
  while(!ready());  //  TODO needed?
  uint8_t buffer[1];
  _readRegister(buffer, 1);
  uint8_t value = (buffer[0] >> 1) & 0x03;
  return value;
}


//  PAGE 22 - experimental
//  DAC value is reset to EEPROM value
//  need to reflect this in cached value
int MCP4725::powerOnReset()
{
  int rv = _generalCall(MCP4725_GC_RESET);
  _lastValue = readDAC(); //  update cache to actual value;
  return rv;
}


//  PAGE 22 - experimental
//  _powerDownMode DAC resets to 0 -- PDM EEPROM stays same !!!
//  need to reflect this in cached value
int MCP4725::powerOnWakeUp()
{
  int rv = _generalCall(MCP4725_GC_WAKEUP);
  _powerDownMode = readPowerDownModeDAC();  //  update to actual value;
  return rv;
}


//  PAGE 18 DATASHEET
int MCP4725::_writeFastMode(const uint16_t value)
{
  uint8_t l = value & 0xFF;
  uint8_t h = ((value / 256) & 0x0F);  //  set C0 = C1 = 0, no PDmode
  h = h | (_powerDownMode << 4);

  _wire->beginTransmission(_deviceAddress);
  _wire->write(h);
  _wire->write(l);
  return _wire->endTransmission();
}


//  PAGE 19 DATASHEET
//  reg = MCP4725_DAC | MCP4725_EEPROM
int MCP4725::_writeRegisterMode(const uint16_t value, uint8_t reg)
{
  if (reg & MCP4725_DACEEPROM)
  {
    _lastWriteEEPROM = millis();
  }
  uint8_t h = (value / 16);
  uint8_t l = (value & 0x0F) << 4;
  _wire->beginTransmission(_deviceAddress);
  reg = reg | (_powerDownMode << 1);
  _wire->write(reg);
  _wire->write(h);
  _wire->write(l);
  return _wire->endTransmission();
}


//  PAGE 20 DATASHEET
//  typical 3 or 5 bytes
uint8_t MCP4725::_readRegister(uint8_t* buffer, const uint8_t length)
{
  _wire->beginTransmission(_deviceAddress);
  int rv = _wire->endTransmission();
  if (rv != 0) return 0;  //  error

  //  readBytes will always be equal or smaller to length
  uint8_t readBytes = _wire->requestFrom(_deviceAddress, length);
  uint8_t cnt = 0;
  while (cnt < readBytes)
  {
    buffer[cnt++] = _wire->read();
  }
  return readBytes;
}


//  name comes from datasheet
int MCP4725::_generalCall(const uint8_t gc)
{
  _wire->beginTransmission(0);  //  _deviceAddress
  _wire->write(gc);
  return _wire->endTransmission();
}


//  -- END OF FILE --



MCP4725.h
#pragma once
//
//    FILE: MCP4725.h
//  AUTHOR: Rob Tillaart
// PURPOSE: Arduino library for 12 bit I2C DAC - MCP4725 
// VERSION: 0.3.3
//     URL: https://github.com/RobTillaart/MCP4725
//


#include "Wire.h"
#include "Arduino.h"


#define MCP4725_VERSION         (F("0.3.3"))


// constants
#define MCP4725_MAXVALUE        4095


// errors
#define MCP4725_OK              0
#define MCP4725_VALUE_ERROR     -999
#define MCP4725_REG_ERROR       -998
#define MCP4725_NOT_CONNECTED   -997


// powerDown Mode - TODO ENUM?
#define MCP4725_PDMODE_NORMAL   0x00
#define MCP4725_PDMODE_1K       0x01
#define MCP4725_PDMODE_100K     0x02
#define MCP4725_PDMODE_500K     0x03


class MCP4725
{
public:
  explicit MCP4725(const uint8_t deviceAddress, TwoWire *wire = &Wire);

#if defined(ESP8266) || defined(ESP32)
  bool     begin(const uint8_t dataPin, const uint8_t clockPin);
#endif

#if defined (ARDUINO_ARCH_RP2040)

  bool    begin(int sda, int scl);

#endif

  bool     begin();
  bool     isConnected();


  // uses writeFastMode
  int      setValue(const uint16_t value = 0);
  // returns last value set - cached - much faster than readDAC();
  uint16_t getValue();


  // 0..100.0% - no input check.
  int      setPercentage(float percentage = 0);
  float    getPercentage() { return getValue() * (100.0 / MCP4725_MAXVALUE); };


  int      writeDAC(const uint16_t value, const bool EEPROM = false);
  // RDY is depreciated in the future, use ready() instead.
  // inline bool RDY()  { return ready(); };
  bool     ready();
  uint32_t getLastWriteEEPROM()  { return _lastWriteEEPROM; };
  uint16_t readDAC();
  uint16_t readEEPROM();


  // experimental
  int      writePowerDownMode(const uint8_t PDM, const bool EEPROM = false);
  uint8_t  readPowerDownModeEEPROM();
  uint8_t  readPowerDownModeDAC();
  int      powerOnReset();
  int      powerOnWakeUp();


private:
  uint8_t  _deviceAddress;
  uint16_t _lastValue;
  uint8_t  _powerDownMode;      // DATASHEET P15?
  int      _writeFastMode(const uint16_t value);
  uint32_t _lastWriteEEPROM;

  int      _writeRegisterMode(const uint16_t value, uint8_t reg);
  uint8_t  _readRegister(uint8_t* buffer, const uint8_t length);

  int      _generalCall(const uint8_t gc);

  TwoWire*  _wire;
};


// -- END OF FILE --

